shell/platform/android/platform_view_android_jni_impl.cc

```diff
@@ -145,6 +146,84 @@ static void DestroyJNI(JNIEnv* env, jobject jcaller, jlong shell_holder) {
   delete ANDROID_SHELL_HOLDER;
 }
 
+// Signature is similar to RunBundleAndSnapshotFromLibrary but it can't change
+// the bundle path or asset manager since we can only spawn with the same
+// AOT.
+//
+// The shell_holder instance must be a pointer address to the current
+// AndroidShellHolder whose Shell will be used to spawn a new Shell.
+//
+// This creates a Java Long that points to the newly created
+// AndroidShellHolder's raw pointer, connects that Long to a newly created
+// FlutterJNI instance, then returns the FlutterJNI instance.
+static jobject SpawnJNI(JNIEnv* env,
+                        jobject jcaller,
+                        jlong shell_holder,
+                        jstring jEntrypoint,
+                        jstring jLibraryUrl) {
+  jclass jniClass = env->FindClass("io/flutter/embedding/engine/FlutterJNI");
+  if (jniClass == nullptr) {
+    FML_LOG(ERROR) << "Could not locate FlutterJNI class";
+    return nullptr;
+  }
+  jmethodID jniConstructor = env->GetMethodID(jniClass, "<init>", "()V");
+  if (jniConstructor == nullptr) {
+    FML_LOG(ERROR) << "Could not locate FlutterJNI's constructor";
+    return nullptr;
+  }
+
+  jobject jni = env->NewObject(jniClass, jniConstructor);
+  if (jni == nullptr) {
+    FML_LOG(ERROR) << "Could not create a FlutterJNI instance";
+    return nullptr;
+  }
+
+  fml::jni::JavaObjectWeakGlobalRef java_jni(env, jni);
+  std::shared_ptr<PlatformViewAndroidJNI> jni_facade =
+      std::make_shared<PlatformViewAndroidJNIImpl>(java_jni);
+
+  auto entrypoint = fml::jni::JavaStringToString(env, jEntrypoint);
+  auto libraryUrl = fml::jni::JavaStringToString(env, jLibraryUrl);
+
+  auto spawned_shell_holder =
+      ANDROID_SHELL_HOLDER->Spawn(jni_facade, entrypoint, libraryUrl);
+
+  if (spawned_shell_holder == nullptr || !spawned_shell_holder->IsValid()) {
+    FML_LOG(ERROR) << "Could not spawn Shell";
+    return nullptr;
+  }
+
+  jfieldID shellHolderId =
+      env->GetFieldID(jniClass, "nativeShellHolderId", "Ljava/lang/Long;");
```

Comment:
Not a huge deal, but this would be faster if we cached the FieldIds and MethodIds.   Maybe it can be a follow up PR.

Reply:
aha, good point. Done